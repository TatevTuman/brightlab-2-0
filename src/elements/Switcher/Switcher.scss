@import 'static/assets/styles/variables';

.switcher {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 312px;
  height: 40px;
  border: 0.5px solid var(--secondary);
  border-radius: 30px;
  cursor: pointer;

  &-section {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    width: 104px;
    border-radius: 30px;
    font-size: 14px;
    font-weight: bold;
    color: #1D2069;

    &:hover {
      background: var(--white-p);
    }
    &:active {
      background: rgba(62, 66, 223, 0.09);
    }

    &[data-active='true'] {
      background: var(--purple);
      color: var(--white);
    }
  }
}