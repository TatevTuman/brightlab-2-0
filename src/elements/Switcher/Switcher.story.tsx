import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import Switcher, { SwitcherProps } from './Switcher'
import { storybook } from '@utils'
import { action } from '@storybook/addon-actions'

interface SwitcherStoryTemplateProps extends SwitcherProps {}

const options = [
  {
    label: 'Month',
    value: 'month'
  },
  {
    label: 'Day',
    value: 'day'
  },
  {
    label: 'Week',
    value: 'week'
  }
]

export default {
  title: 'Elements/Switcher',
  component: Switcher,
  argTypes: {
    value: { control: storybook.controls.select(options.map(option => option.value)), name: 'Value' },
    onCLick: { action: 'onClick' }
  }
} as Meta

const SwitcherStoryTemplate: Story<SwitcherStoryTemplateProps> = args => {
  return <Switcher {...args} options={options} />
}

SwitcherStoryTemplate.args = {
  value: options[0].value,
  onClick: action('onClick')
}

export const SwitcherStory = SwitcherStoryTemplate.bind({})
SwitcherStory.args = {
  ...SwitcherStoryTemplate.args
}
