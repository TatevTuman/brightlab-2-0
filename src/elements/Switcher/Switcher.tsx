import React, { memo } from 'react'
import { isEqual } from 'lodash'
import { OptionType } from '@types'
import './Switcher.scss'

export interface SwitcherProps {
  options: OptionType[]
  value: string
  onClick(value: string): void
}

const Switcher: React.FC<SwitcherProps> = props => {
  const { options, value, onClick } = props

  let side = null

  options.forEach((option, index) => {
    if (isEqual(option.value, value)) {
      side = index === 0 ? (side = 'left') : index == 1 ? (side = 'center') : (side = 'right')
    }
  })

  return (
    <div className={'switcher'}>
      <div className={'switcher-side'} data-side={side} />
      {options.map(option => {
        const { label } = option
        const isActive = isEqual(option.value, value)

        return (
          <div key={label} className={'switcher-section'} onClick={() => onClick(option.value)} data-active={isActive}>
            <span>{label}</span>
          </div>
        )
      })}
    </div>
  )
}
Switcher.defaultProps = {
  value: 'week',
  options: [
    {
      label: 'Month',
      value: 'month'
    },
    {
      label: 'Day',
      value: 'day'
    },
    {
      label: 'Week',
      value: 'week'
    }
  ]
}

export default memo(Switcher)
