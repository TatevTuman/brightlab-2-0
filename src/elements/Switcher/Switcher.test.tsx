import React from 'react'
import { render, cleanup, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { OptionType } from '@types'

import Switcher from './Switcher'

afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Switcher', () => {
  const options: OptionType[] = new Array(3).fill(0).map((_, index) => {
    return {
      label: 'test-' + (index + 1),
      value: (index + 1).toString()
    }
  })

  const props = {
    options,
    value: '1',
    onClick: jest.fn()
  }

  it('renders correctly with props', async () => {
    const { container } = render(<Switcher {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('renders correctly label and value', async () => {
    const { getByText } = render(<Switcher {...props} />)

    options.forEach(({ label }) => {
      expect(getByText(label)).toBeInTheDocument()
    })
  })

  it('data-active have current value if the props value matches the value of the first switch', async () => {
    const { getByText } = render(<Switcher {...props} />)

    const firstSwitch = getByText('test-1') as HTMLElement
    expect(firstSwitch).toHaveAttribute('data-active', 'true')

    const secondSwitch = getByText('test-2') as HTMLElement
    expect(secondSwitch).toHaveAttribute('data-active', 'false')

    const thirdSwitch = getByText('test-3') as HTMLElement
    expect(thirdSwitch).toHaveAttribute('data-active', 'false')
  })

  it('click', async () => {
    const { getByText } = render(<Switcher {...props} />)
    const firstSwitch = getByText('test-1') as HTMLElement

    expect(firstSwitch).toHaveAttribute('data-active', 'true')

    act(() => {
      userEvent.click(firstSwitch!)
    })

    expect(props.onClick).toHaveBeenCalledTimes(1)
  })
})
