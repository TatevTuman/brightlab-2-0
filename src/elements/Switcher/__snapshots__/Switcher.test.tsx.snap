// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switcher renders correctly with props 1`] = `
<div>
  <div
    class="switcher"
  >
    <div
      class="switcher-side"
      data-side="left"
    />
    <div
      class="switcher-section"
      data-active="true"
    >
      <span>
        test-1
      </span>
    </div>
    <div
      class="switcher-section"
      data-active="false"
    >
      <span>
        test-2
      </span>
    </div>
    <div
      class="switcher-section"
      data-active="false"
    >
      <span>
        test-3
      </span>
    </div>
  </div>
</div>
`;
